src/main/java/ch/njol/skript/bukkitutil/ComponentWrapper.java

```diff
@@ -0,0 +1,92 @@
+package ch.njol.skript.bukkitutil;
+
+import ch.njol.skript.util.ItemSource;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.BiConsumer;
+
+/**
+ * Used for storing the component and/or the item.
+ * @param <T> The type of component
+ */
+public abstract class ComponentWrapper<T> {
+
+	private @Nullable ItemSource itemSource;
+	private @Nullable ItemStack itemStack;
+	private final T component;
+
+	/**
+	 * Create a {@link ComponentWrapper} that converts {@code itemStack} into a {@link ItemSource}
+	 * @param itemStack The original {@link ItemStack}
+	 */
+	public ComponentWrapper(ItemStack itemStack) {
+		this(new ItemSource(itemStack));
+	}
+
+	public ComponentWrapper(ItemSource itemSource) {
+		this.itemSource = itemSource;
+		this.itemStack = itemSource.getItemStack();
+		this.component = convertItemStack(itemStack);
+	}
+
+	/**
+	 * Create a {@link ComponentWrapper} that only references to a component
+	 */
+	public ComponentWrapper(T component) {
+		this.component = component;
+	}
+
+	public T getComponent() {
+		return component;
+	}
+
+	public @Nullable ItemStack getItemStack() {
+		return itemStack;
+	}
+
+	public @Nullable ItemSource getItemSource() {
+		return itemSource;
+	}
+
+	/**
+	 * Convert the {@link ItemStack} to the component type.
+	 */
+	public abstract T convertItemStack(ItemStack itemStack);
```

Comment:
this method confuses me. If it's irrespective of the wrapper's state and just converts the input, it should be static. If it's dependent on state, it should just use the itemstack of the wrapper itself. Should it even be public?